crypto/bls_thresholdsign.go

```diff
@@ -0,0 +1,574 @@
+// +build relic
+
+package crypto
+
+// #cgo CFLAGS: -g -Wall -std=c99
+// #include "bls_thresholdsign_include.h"
+import "C"
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/onflow/flow-go/crypto/hash"
+)
+
+// BLS-based threshold signature on BLS 12-381 curve
+// The BLS settings are the same as in the signature
+// scheme defined in the package.
+
+// A threshold signature scheme allows any subset of (t+1)
+// valid signature shares to reconstruct the threshold signature.
+// Up to (t) shares do not reveal any information about the threshold
+// signature.
+// Although the API allows using arbitrary values of (t),
+// the threshold signature scheme is secure in the presence of up to (t)
+// malicious participants when (t < n/2).
+// In order to optimize equally for unforgeability and robustness,
+// the input threshold value (t) should be set to t = floor((n-1)/2).
+
+// The package offers two api for BLS threshold signature:
+// - stateful api where a structure holds all information
+//  of the threshold signature protocols and is recommended
+//  to be used for safety and to reduce protocol inconsistencies.
+// - stateless api with signature reconstruction. Verifying and storing
+// the signature shares has to be managed outside of the library.
+
+// blsThresholdSignatureInspector implements ThresholdSignatureInspector
+// based on the BLS signature scheme
+type blsThresholdSignatureInspector struct {
+	// size of the group
+	size int
+	// the threshold t of the scheme where (t+1) shares are
+	// required to reconstruct a signature
+	threshold int
+	// the group public key (a threshold KG output)
+	groupPublicKey PublicKey
+	// the group public key shares (a threshold KG output)
+	publicKeyShares []PublicKey
+	// the hasher to be used for all signatures
+	hasher hash.Hasher
+	// the message to be signed. Signature shares and the threshold signature
+	// are verified against this message
+	message []byte
+	// the valid signature shares received from other participants
+	shares map[index]Signature
+	// the threshold signature. It is equal to nil if less than (t+1) shares are
+	// received
+	thresholdSignature Signature
+	// lock for atomic operations
+	lock sync.RWMutex
+}
+
+// blsThresholdSignatureParticipant implements ThresholdSignatureParticipant
```

Comment:
The `blsThresholdSignatureParticipant`'s feels somewhat mixed in with the `blsThresholdSignatureInspector`. Just reading over the code it makes me jump back and forth between the different objects. Maybe we could move the `blsThresholdSignatureInspector` to its own file? Or maybe have one first struct and its methods, then the other struct followed by its methods? Just a stylistic preference ... fee free to ignore.

Reply:
I did some rearrangement of functions/structs but I left the 2 structs in the same file as are tightly related.

`blsThresholdSignatureParticipant` is the main struct in the file (although flow consensus does not use it). and `blsThresholdSignatureInspector` is its main component. Both struct share most of the methods as a consequence of embedding.